core/base/v7/inc/ROOT/TDirectoryEntry.h

```diff
@@ -26,77 +26,70 @@ namespace Experimental {
-  TClass* fType = 0; ///< The type of the object
-
-protected:
-  TDirectoryEntryPtrBase() = default;
-
 public:
-  /// Constructor setting the type.
-  TDirectoryEntryPtrBase(TClass* cl): fType(cl) {}
+  TDirectoryEntry(): TDirectoryEntry(nullptr) {}
+  TDirectoryEntry(std::nullptr_t):
+    TDirectoryEntry(std::make_shared<std::nullptr_t>(nullptr)) {}
+  template<class T>
+  explicit TDirectoryEntry(T* ptr):
+    TDirectoryEntry(std::make_shared<T>(*ptr)) {}
+  template<class T>
+  explicit TDirectoryEntry(const std::shared_ptr<T>& ptr):
+    fTypeInfo(&typeid(T)),
+    fType(TClass::GetClass(typeid(T))),
+    fObj(ptr) {}
 
   /// Get the last change date of the entry.
   const time_point_t& GetDate() const { return fDate; }
 
-  /// Get the object's type.
-  TClass* GetType() const { return fType; }
-
   /// Inform the entry that it has been modified, and needs to update its
   /// last-changed time stamp.
   void SetChanged() { fDate = clock_t::now(); }
 
-  /// Abstract interface to retrieve the address of the object represented by
-  /// this entry, if any.
-  virtual void* GetObjectAddr() const = 0;
-
-  /// Abstract interface to retrieve the type of the object represented by this
-  /// entry.
-  virtual const std::type_info& GetTypeInfo() const = 0;
-};
-
-
-/**
- Type-aware part of an entry in a TDirectory. It can manage the lifetime of an
- object, or only inform about last-change date and the type of an object.
- */
-
-template <class T>
-class TDirectoryEntryPtr: public TDirectoryEntryPtrBase {
-public:
-  /// Initialize a TDirectoryEntryPtr from an existing object ("write").
-  TDirectoryEntryPtr(T&& obj):
-    TDirectoryEntryPtrBase(TClass::GetClass(typeid(T))),  fObj(obj) {}
-
-  /// Initialize a TDirectoryEntryPtr from an existing object ("write").
-  TDirectoryEntryPtr(const std::shared_ptr<T>& ptr): fObj(ptr) {}
+  /// Type of the object represented by this entry.
+  const std::type_info& GetTypeInfo() const { return *fTypeInfo; }
 
-  virtual ~TDirectoryEntryPtr() {}
+  /// Get the object's type.
+  TClass* GetType() const { return fType; }
 
   /// Retrieve the `shared_ptr` of the referenced object.
-  std::shared_ptr<T>& GetPointer() const { return fObj; }
+  std::shared_ptr<void>& GetPointer() { return fObj; }
+  const std::shared_ptr<void>& GetPointer() const { return fObj; }
 
-  /// Retrieve the address of the object managed by this entry.
-  /// Returns `nullptr` if no object is managed by this entry.
-  void* GetObjectAddr() const final { return fObj.get(); }
+  template<class U>
+  std::shared_ptr<U> CastPointer() const;
 
-  /// Type of the object represented by this entry.
-  const std::type_info& GetTypeInfo() const final { return typeid(T); }
+  explicit operator bool() const { return *fTypeInfo != typeid(std::nullptr_t); }
 
 private:
-  std::shared_ptr<T> fObj; ///< Object referenced by this entry, if any.
+  time_point_t fDate = clock_t::now(); ///< Time of last change
```

Comment:
I am not sure what you are referring to.  We have many developers, each using their own development tools and their favorite editors.   In addition, most of them are working on different projects (which usually have different favorite indentation size).   This results in each having a different setting for the 'number of spaces' the editor replaces tab with ... resulting in misaligned code.

If something has happened that relief this situation and make the tab be replaced by the correct number of spaces no matter what, I am not aware of it. (In addition, I am not aware on any advantages of using tab characters).

As a side note I can send you a work-in-progress .clang-format that is close to what is needed (but not quite, in part due to quirk in clang-format).

thanks.